In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine
import psycopg2


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
def queryResults (q):
    return pd.io.sql.read_sql(q, con=engine)

## Describe the goals of the study.
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

### Task
- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating: look only at books with at least 50 ratings.
- Find the average number of text reviews among users who rated more than 50 books.


### Instructions for completing the task
- Describe the goals of the study.
- Study the tables (print the first rows).
- Make an SQL query for each of the tasks.
- Output the results of each query in the Notebook.
- Describe your conclusions for each of the tasks.

## Study the tables (print the first rows).


In [3]:
query = '''

SELECT 
* 
FROM 
books 
LIMIT 10;
'''
queryResults(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
query = '''

SELECT 
* 
FROM 
authors 
LIMIT 10;
'''
queryResults(query)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [5]:
query = '''

SELECT 
* 
FROM 
publishers 
LIMIT 10;
'''
queryResults(query)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [6]:
query = '''

SELECT 
* 
FROM 
ratings 
LIMIT 10;
'''
queryResults(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [7]:
query = '''

SELECT 
* 
FROM 
reviews
LIMIT 10;
'''
queryResults(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## SQL query for each of the tasks

### Find the number of books released after January 1, 2000.


In [8]:
query = '''

SELECT 
COUNT(DISTINCT books.book_id) AS book_cnt
FROM 
books
WHERE
publication_date >= '2000-01-01'

'''
queryResults(query)

book_cnt
0       821

821 books were released after January 1, 2000.

### Find the number of user reviews and the average rating for each book.


In [9]:
query = '''

SELECT
books.title,
COUNT(reviews.review_id) AS review_cnt,
AVG(ratings.rating) AS avg_rating
FROM
reviews
INNER JOIN books ON reviews.book_id = books.book_id
INNER JOIN ratings ON reviews.book_id = ratings.book_id
GROUP BY
books.title
ORDER BY
avg_rating DESC, review_cnt DESC
LIMIT 60

'''
queryResults(query)

title  review_cnt  avg_rating
0                       A Dirty Job (Grim Reaper  #1)          16    5.000000
1             School's Out—Forever (Maximum Ride  #2)          12    5.000000
2        Moneyball: The Art of Winning an Unfair Game           9    5.000000
3   In the Hand of the Goddess (Song of the Liones...           6    5.000000
4                                My Name Is Asher Lev           4    5.000000
5                   Misty of Chincoteague (Misty  #1)           4    5.000000
6   The Ghost Map: The Story of London's Most Terr...           4    5.000000
7   The War of Art: Break Through the Blocks & Win...           4    5.000000
8            Evil Under the Sun (Hercule Poirot  #24)           4    5.000000
9                                               March           4    5.000000
10               Stolen (Women of the Otherworld  #2)           4    5.000000
11                       Plum Island (John Corey  #1)           4    5.000000
12                                 The Woman in Black           4    5.000000
13                           Neil Gaiman's Neverwhere           4    5.000000
14               Welcome to Temptation (Dempseys  #1)           4    5.000000
15  The Walking Dead  Book One (The Walking Dead #...           4    5.000000
16              A Fistful of Charms (The Hollows  #4)           4    5.000000
17                                        Angels Fall           4    5.000000
18                                      Alas  Babylon           4    5.000000
19          A Woman of Substance (Emma Harte Saga #1)           4    5.000000
20                                Homage to Catalonia           4    5.000000
21  Captivating: Unveiling the Mystery of a Woman'...           4    5.000000
22               Pop Goes the Weasel (Alex Cross  #5)           4    5.000000
23                                         Hard Times           4    5.000000
24      Arrows of the Queen (Heralds of Valdemar  #1)           4    5.000000
25                           Tai-Pan (Asian Saga  #2)           4    5.000000
26               Piercing the Darkness (Darkness  #2)           4    5.000000
27                                      Evening Class           4    5.000000
28                 The Big Bad Wolf  (Alex Cross  #9)           4    5.000000
29                          Triptych (Will Trent  #1)           4    5.000000
30  Crucial Conversations: Tools for Talking When ...           4    5.000000
31                                          Geek Love           4    5.000000
32  The Demon-Haunted World: Science as a Candle i...           4    5.000000
33                      Emily of New Moon (Emily  #1)           4    5.000000
34                    Act of Treason (Mitch Rapp  #9)           4    5.000000
35                                         Dead Souls           4    5.000000
36  Wherever You Go  There You Are: Mindfulness Me...           4    5.000000
37                                    Light in August           4    5.000000
38                                              Women           4    5.000000
39                                        Marvel 1602           4    5.000000
40  The Adventures of Tom Sawyer and Adventures of...           2    5.000000
41  How to Be a Domestic Goddess: Baking and the A...           2    5.000000
42  Stone of Farewell (Memory  Sorrow  and Thorn  #2)           2    5.000000
43                 Dragonfly in Amber (Outlander  #2)          32    4.875000
44             Preludes & Nocturnes (The Sandman  #1)          18    4.833333
45                                    House of Leaves          16    4.750000
46  The Witching Hour (Lives of the Mayfair Witche...          12    4.750000
47                                          The Twits          12    4.750000
48      Lover Revealed (Black Dagger Brotherhood  #4)          12    4.750000
49                The Monster at the End of this Book          12    4.750000
50          The Amazing Adventures of Kavalier & Clay          24    4

In [10]:
query = '''

SELECT
books.title,
COUNT(reviews.review_id) AS review_cnt,
AVG(ratings.rating) AS avg_rating
FROM
reviews
INNER JOIN books ON reviews.book_id = books.book_id
INNER JOIN ratings ON reviews.book_id = ratings.book_id
GROUP BY
books.title
ORDER BY
review_cnt DESC, avg_rating
LIMIT 10

'''
queryResults(query)

title  review_cnt  avg_rating
0                            Twilight (Twilight  #1)        1120    3.662500
1                The Hobbit  or There and Back Again         528    4.125000
2                             The Catcher in the Rye         516    3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...         492    4.414634
4  Harry Potter and the Chamber of Secrets (Harry...         480    4.287500
5               Angels & Demons (Robert Langdon  #1)         420    3.678571
6  Harry Potter and the Order of the Phoenix (Har...         375    4.186667
7  The Lightning Thief (Percy Jackson and the Oly...         372    4.080645
8                                        Animal Farm         370    3.729730
9  The Fellowship of the Ring (The Lord of the Ri...         370    4.391892

43 books have the highest number (5 points) in the rating category, but only between 2 and 4 number of ratings. The only exception are 'A Dirty Job (Grim Reaper #1)' with 16 number of ratings, 'School's Out—Forever (Maximum Ride #2)' with 12 ratings, 'Moneyball: The Art of Winning an Unfair Game' with 9 ratings and 'In the Hand of the Goddess (Song of the Liones...' with 6 ratings.

The book with the highest number of ratings has the book 'Twilight (Twilight #1)' with an average rating of 3,66 followed by the popular books 'The Hobbit or There and Back Again' (528 ratings) with a higher average rating of 4,13 and 'The Catcher in the Rye' with 516 ratings with 3,83. The following 7 books from different book series except 'Animal farm' hava also a high number of ratings between (370 and 492) and a relativly high average rating (3,68 - 4,41). 

### Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [11]:
query = '''

SELECT
publishers.publisher,
COUNT(DISTINCT books.book_id) AS book_cnt
FROM
publishers
INNER JOIN books ON publishers.publisher_id = books.publisher_id
WHERE
books.num_pages > 50
GROUP BY
publishers.publisher
ORDER BY
COUNT(books.book_id) DESC 
LIMIT 10;

'''
queryResults(query)

publisher  book_cnt
0             Penguin Books        42
1                   Vintage        31
2  Grand Central Publishing        25
3          Penguin Classics        24
4          Ballantine Books        19
5                    Bantam        19
6                   Berkley        17
7             Berkley Books        14
8        St. Martin's Press        14
9                     Delta        13

The publisher with the highest number of released books are Penguin Books with 42 books, followed with a big leap by Vintage with 31 books. Grand Central Publishing releases 25 books closely followed by Penguin Classics (24).

### Identify the author with the highest average book rating: look only at books with at least 50 ratings.


In [12]:
query = '''

SELECT
AVG(ratings.rating) AS avg_rating,
authors.author AS author
FROM
books
LEFT JOIN authors ON books.author_id = authors.author_id
LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY
author
HAVING
COUNT(ratings.rating_id) > 50
ORDER BY
avg_rating DESC
LIMIT 10;

'''
queryResults(query)

avg_rating                             author
0    4.288462         J.K. Rowling/Mary GrandPré
1    4.283019                    Agatha Christie
2    4.264151  Markus Zusak/Cao Xuân Việt Khương
3    4.240964                     J.R.R. Tolkien
4    4.209677           Roald Dahl/Quentin Blake
5    4.203704                  Louisa May Alcott
6    4.130952                       Rick Riordan
7    4.107143                      Arthur Golden
8    4.009434                       Stephen King
9    3.971429                       John Grisham

The autho with the highest average rating is J.K.Rowling & Mary GrandPre with an average rating of 4.29, closely followed by Agatha Christie (4.28) and on the third place you can find following authors Markus Zusak and Cao Xuân Việt Khương with an average rating of 4.26.

### Find the average number of text reviews among users who rated more than 50 books.


In [13]:

query = '''
 
        WITH SUBQ1 AS (
            SELECT
                ratings.username,
                count (ratings.rating_id) as num_of_ratings
            
            FROM
                ratings

            GROUP BY
                ratings.username

            HAVING
                count(ratings.rating_id)> 50
            ), 
        SUBQ2 AS (
            SELECT 
                reviews.username,
                count (reviews.text) as num_of_texts

            FROM
                reviews

            GROUP BY
                reviews.username
            )

        SELECT 
            AVG (SUBQ2.num_of_texts)::integer AS avg_num_review
            
            FROM 
                SUBQ1 JOIN SUBQ2
                ON SUBQ1.username = SUBQ2.username
'''
queryResults(query)

avg_num_review
0              24

The average number of text reviews among user who rated more than 50 books is 24
